File: problems/857_minimum_cost_to_hire_k_workers/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    double mincostToHireWorkers(vector<int>& quality, vector<int>& wage, int k) {
        int n = quality.size();
        vector<pair<double, int>> workers(n);
        for (int i = 0; i < n; i++) workers[i] = {(double)wage[i] / quality[i], quality[i]};
        sort(workers.begin(), workers.end());
        priority_queue<int> pq;
        double result = DBL_MAX;
        int totalQuality = 0;
        for (auto& [ratio, q] : workers) {
            pq.push(q);
            totalQuality += q;
            if ((int)pq.size() > k) { totalQuality -= pq.top(); pq.pop(); }
            if ((int)pq.size() == k) result = min(result, totalQuality * ratio);
        }
        return result;
    }
};

File: problems/857_minimum_cost_to_hire_k_workers/solution.py
from typing import List


class Solution:
    def mincostToHireWorkers(self, quality: List[int], wage: List[int], k: int) -> float:
        workers = sorted((w / q, q) for w, q in zip(wage, quality))
        result = float('inf')
        total_quality = 0
        heap = []
        for ratio, q in workers:
            heapq.heappush(heap, -q)
            total_quality += q
            if len(heap) > k:
                total_quality += heapq.heappop(heap)
            if len(heap) == k:
                result = min(result, total_quality * ratio)
        return result
